pkg/ddc/base/operation.go

```diff
@@ -0,0 +1,197 @@
+/*
+Copyright 2022 The Fluid Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package base
+
+import (
+	"fmt"
+	datav1alpha1 "github.com/fluid-cloudnative/fluid/api/v1alpha1"
+	"github.com/fluid-cloudnative/fluid/pkg/common"
+	"github.com/fluid-cloudnative/fluid/pkg/dataoperation"
+	cruntime "github.com/fluid-cloudnative/fluid/pkg/runtime"
+	"github.com/fluid-cloudnative/fluid/pkg/utils"
+	v1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+func (t *TemplateEngine) Operate(ctx cruntime.ReconcileRequestContext, object client.Object, opStatus *datav1alpha1.OperationStatus,
+	operation dataoperation.OperationInterface) (ctrl.Result, error) {
+	operateType := operation.GetOperationType()
+
+	// runtime engine override the template engine
+	switch operateType {
+	case dataoperation.DataBackup:
+		ownImpl, ok := t.Implement.(Databackuper)
+		if ok {
+			targetDataBackup, success := object.(*datav1alpha1.DataBackup)
+			if !success {
+				return utils.RequeueIfError(fmt.Errorf("object %v is not a DataBackup", object))
+			}
+			return ownImpl.BackupData(ctx, *targetDataBackup)
+		}
+	}
+
+	// use default template engine
+	switch opStatus.Phase {
+	case common.PhaseNone:
+		return t.reconcileNone(ctx, object, opStatus, operation)
+	case common.PhasePending:
+		return t.reconcilePending(ctx, object, opStatus, operation)
+	case common.PhaseExecuting:
+		return t.reconcileExecuting(ctx, object, opStatus, operation)
+	case common.PhaseComplete:
+		return t.reconcileComplete(ctx, object, opStatus, operation)
+	case common.PhaseFailed:
+		return t.reconcileFailed(ctx, object, operation)
+	default:
+		ctx.Log.Error(fmt.Errorf("unknown phase"), "won't reconcile it", "phase", opStatus.Phase)
+		return utils.NoRequeue()
+	}
+}
+
+func (t *TemplateEngine) reconcileNone(ctx cruntime.ReconcileRequestContext, object client.Object,
+	opStatus *datav1alpha1.OperationStatus, operation dataoperation.OperationInterface) (ctrl.Result, error) {
+	log := ctx.Log.WithName("reconcileNone")
+
+	// 0. check the object spec valid or not
+	conditions, err := operation.Validate(ctx, object)
+	if err != nil {
+		log.Error(err, "validate failed")
```

Comment:
Suggest to add namespace and name of the object in `log.Error`

Reply:
added.